app/models/annotation_category.rb

```diff
@@ -23,62 +23,64 @@ class AnnotationCategory < ApplicationRecord
-        )
-        unless annotation_text.save
-          raise CsvInvalidLineError, I18n.t('annotation_categories.upload.error',
-                                            annotation_category: annotation_category.annotation_category_name)
+    AnnotationCategory.transaction do
+      name = row.shift
+      annotation_category = assignment.annotation_categories.find_by(annotation_category_name: name)
+      # The second column is the optional flexible criterion name.
+      criterion_name = row.shift
+      if annotation_category.nil?
+        annotation_category = assignment.annotation_categories.create(annotation_category_name: name)
+        unless annotation_category.valid?
+          raise CsvInvalidLineError, I18n.t('annotation_categories.upload.empty_category_name')
         end
+      elsif (annotation_category.flexible_criterion_id.nil? && !criterion_name.nil?) ||
+          (annotation_category.flexible_criterion.name != criterion_name)
+        raise CsvInvalidLineError, I18n.t('annotation_categories.upload.invalid_criterion',
+                                          annotation_category: name)
       end
-    else
-      criterion = assignment.criteria.find_by(name: criterion_name, type: 'FlexibleCriterion')
-      if criterion.nil?
-        raise CsvInvalidLineError, I18n.t('annotation_categories.upload.criterion_not_found',
-                                          missing_criterion: criterion_name)
-      end
-      annotation_category.update!(flexible_criterion_id: criterion.id)
-      row.each_slice(2) do |text_with_deduction|
-        begin
-          new_deduction = Float(text_with_deduction.second)
-        rescue ArgumentError, TypeError
-          raise CsvInvalidLineError, I18n.t('annotation_categories.upload.deduction_absent',
-                                            value: text_with_deduction.second,
-                                            annotation_category: annotation_category.annotation_category_name)
+      if criterion_name.nil?
+        row.each do |text|
+          annotation_text = annotation_category.annotation_texts.build(
+            content: text,
+            creator_id: current_user.id,
+            last_editor_id: current_user.id
+          )
+          unless annotation_text.save
+            raise CsvInvalidLineError, I18n.t('annotation_categories.upload.error',
+                                              annotation_category: annotation_category.annotation_category_name)
+          end
         end
-        if new_deduction > criterion.max_mark || new_deduction < 0
-          raise CsvInvalidLineError, I18n.t('annotation_categories.upload.invalid_deduction',
-                                            annotation_content: text_with_deduction.first,
-                                            criterion_name: criterion_name,
-                                            value: new_deduction)
+      else
+        criterion = assignment.flexible_criteria.find_by(name: criterion_name)
```

Comment:
I think the STI for criteria is messing with this, but I can't remember if I am supposed to have the type or not now. @david-yz-liu

Reply:
(oh I rebased off master so you wouldn't have to deal with the merge conflict, so I guess I should check master?)